README.md
# lsystems

### [API docs](https://joebentley.github.io/lsystems/)

[![Clojars Project](https://img.shields.io/clojars/v/lsystems.svg)](https://clojars.org/lsystems)

Simple and easy to use implementation of L-systems in Clojure.

L-systems were invented by Aristid Lindenmayer as a way to model bacteria growth,
and are described as a recursive transformation of a sequence of symbols called a "string".
By taking an initial string and repeatedly applying the transformation rules, called
"productions", complicated evolving strings can be created.

By using [turtle graphics](https://en.wikipedia.org/wiki/Turtle_graphics)–where a pen
is controlled by movement commands and lines are drawn out along its movement path–we
can assign different commands to different symbols, for example:

* `F` could mean "move the pen forward by 10 pixels along its current direction"
* `+` could mean "rotate the pen's direction clockwise 25 degrees"

The symbols will often have a different meaning depending on the specific system. Two
symbols that are usually the same among L-systems are:

* `[` meaning "push the pen's position and direction to the stack"
* `]` meaning "pop the position and direction from the stack
and set the pen's position and direction to that value"

These commands are used to make branching and repeating tree-like structures.

We use the excellent [Clojure2D](https://github.com/Clojure2D/clojure2d) for rendering.

## Examples

To list possible examples:

```
> lein run -m lsystems.examples --list
examples:
fractal-binary-tree
fractal-plant
```

Then to run an example:

```
> lein run -m lsystems.examples fractal-binary-tree
running example fractal-binary-tree...
(window will display)
```

To run all examples pass `--all`.

See [the examples](./src/lsystems/examples.clj) file. I'll explain one here:

```clojure
(require
  [lsystems.core :refer [nth-step]]
  [lsystems.turtle :refer :all]
  [clojure2d.core :as c])

(defn fractal-plant []
  (let [width 600 height 600]
;; 1.
    (render-to-canvas-by-executing-state-with-rules
;; 2.
      (nth-step {\X (seq "F+[[X]-X]-F[-FX]+X") \F '(\F \F)}
                \X
                6)
;; 3.
      {\F (fn [s] (forward s 10))
       \- (fn [s] (rotate s (- 25)))
       \+ (fn [s] (rotate s (+ 25)))
       \[ (fn [s] (push-pos-and-angle s))
       \] (fn [s] (pop-pos-and-angle s))}
;; 4.
      (fn [canvas]
        (show-window canvas "Fractal plant")
        (save-canvas-to-file canvas "example-renders/fractal-plant.jpg"))
;; 5.
      :facing -25 :width width :height height :auto-resize-padding 50
;; 6.
      :canvas-function (fn [canvas]
                         (c/gradient-mode canvas 0 0 :white 0 height :light-green)
                         (c/rect canvas 0 0 width height)
                         (c/set-color canvas :forest-green)
                         (c/set-stroke canvas 1.3)))))

(fractal-plant)
```

1. The call `(render-to-canvas-by-executing-state-with-rules ...)` takes three mandatory arguments,
 a produced L-system state, a series of rules for symbols in the state that take
 the pen state (see [turtle.clj](./src/lsystems/turtle.clj))
 and performs a pen movement action using it (e.g. [`(forward ...)`](https://joebentley.github.io/lsystems/lsystems.turtle.html#var-forward))
 and finally a function that takes the resulting canvas and uses it, e.g. to save it to an image file,
 and optionally many keyword options that can alter the appearance of the produced image. 

2. The function `(nth-step [productions state n])` from [`core.clj`](./src/lsystems/core.clj)
 takes a set of production rules, in this case `{\X (seq "F+[[X]-X]-F[-FX]+X") \F '(\F \F)}`.
 Here we have a map that replaces the character `\X` with the list of characters
 `(seq "F+[[X]-X]-F[-FX]+X")` (**note** you must turn it into a sequence of characters,
 rather than just using a string), while `\F` is replaced by the list `'(\F \F)` (which is the
 same as `(seq "FF")`). `state` is the initial state of the L-system, and `n` is the number
 of iterations to calculate, in this case 6 iterations.
 
3. This is a map from characters in the string to functions that take the current pen state
 and return an updated pen state. For example `\F` moves the pen forward by 10 pixels,
 and `\[` pushes the current pen position and direction to `(pen-state :stack)`.
 See also: [`(standard-rule-set ...)`](https://joebentley.github.io/lsystems/lsystems.turtle.html#var-standard-rule-set).

4. This function uses the resulting canvas created by executing the L-system with the given rules. Here
 we show the canvas in a window with title "Fractal plant" and also save it as a jpeg.

5. Some examples of the possible keyword arguments. `:facing` is the initial pen direction,
`:width` and `:height` are the image dimensions,
 auto resizing of the generated line segments to fit the image is performed by default, and
 `:auto-resize-padding` is used to set the padding around the image.
 
6. A function `:canvas-function` can be passed that takes a
 [Clojure2D Canvas object](https://clojure2d.github.io/clojure2d/docs/codox/clojure2d.core.html)
 which can be used to do some set up _before_ the lines are drawn. For example here we set a gradient
 from white to light green and draw a rect for the background, then we set the stroke
 colour and width for the lines. 

`render-to-canvas-by-executing-state-with-rules` uses `(execute-state-with-rules [state rules pen-state])`
to execute the rules. The pen-state stores the beginning and end of the line segments
internally when `(forward [pen-state by-pixels])` is called, then `(draw-lines [canvas lines])`
is used to draw all the line segments.

The `(forward [pen-state by-pixels])` function actually checks whether the
pen is in the same direction as the previously call by using a cached value stored on `pen-state`, if it does then the
previous line segment is extended rather than a new line being drawn. This helps with performance (less lines to draw)
and also helps with the smoothness of the resulting lines.

Produced image:

![Fractal plant](./example-renders/fractal-plant.jpg)

## TODO

* Stochastic/context-sensitive L-systems
* Write tests for `lsystems.turtle`
* Performance profiling
* Fix centering of images when using `(fit-line-segments-to-screen ...)`, currently they are off centre sometimes
* Program takes a long time to close, maybe unavoidable and it's just the JVM
* Testing on other computers, distribution packaging
* Eventually: 3D, SVG export, user interface

## Directory structure

* `/src` source files
* `/test` tests
* `/doc` markdown files for building docs (build docs with `lein codox`)
* `/docs` codox HTML output directory

## References

* https://jsantell.com/l-systems

* https://en.wikipedia.org/wiki/L-system

* [The Algorithmic Beauty of Plants](http://algorithmicbotany.org/papers/#abop)

## Other examples

### Dragon curve

![Dragon curve](./example-renders/dragon-curve.jpg)

### Fractal Binary Tree

![Fractal binary tree](./example-renders/fractal-binary-tree.jpg)

### Fig 1.24 http://algorithmicbotany.org/papers/abop/abop-ch1.pdf

![Algorithmic Beauty of Plants example](./example-renders/abop-fig1.24.jpg)

## License

Copyright © 2019 Joe Bentley joebentley10@gmail.com

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-12-12
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-12-12
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/lsystems/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/lsystems/compare/0.1.0...0.1.1


test/lsystems/core_test.clj
(ns lsystems.core-test
  (:require [clojure.test :refer :all]
            [lsystems.core :refer :all]))


(deftest test-combine-keywords
  (testing "Return nil if called with no args"
    (is (nil? (combine-keywords))))
  (testing "Return given keyword if called with one arg"
    (is (= (combine-keywords :a) :a)))
  (testing "Combine two properly"
    (is (= (combine-keywords :a :b) :ab)))
  (testing "Combine three (with third having two chars) properly"
    (is (= (combine-keywords :a :b :cd) :abcd))))

(deftest test-string-to-keywords
  (testing "Return empty list on empty string"
    (is (= (string-to-keywords "") '())))
  (testing "Returns expected list of strings"
    (is (= (string-to-keywords "abcd") '(:a :b :c :d)))))

(deftest test-state-to-string
  (testing "Return empty string on empty list"
    (is (= (state-to-string '()) "")))
  (testing "Returns expected string from list, combining all characters"
    (is (= (state-to-string '(:a :b :cd)) "abcd")))
  (testing "Should handle strings and numbers okay"
    (is (= (state-to-string '(:a :b "cd" :z 2)) "abcdz2"))))

(deftest test-adding-production
  (testing "Should add new production to map"
    (is (= (add-production new-productions :a '(:a :b)) {:a '(:a :b)}))))

(deftest test-step
  (let [productions (add-production new-productions :a '(:a :b))]
    (testing "Advance the state as expected, and work when given a keyword rather than list"
      (is (= (step productions :a) [:a :b])))
    (testing "Step using threading operator"
      (is (= (->> :a
                  (step productions)
                  (step productions))
             [:a :b :b]))))
  ; https://en.wikipedia.org/wiki/L-system#Examples_of_L-systems
  (testing "7th step of Algae"
    (let [algae-system {:A [:A :B] :B :A}]
      (is (= (state-to-string (nth-step algae-system :A 7))
             "ABAABABAABAABABAABABAABAABABAABAAB"))))
  (testing "3rd step of Fractal (binary) tree"              ;; also tests using numbers and strings as variables
    (let [binary-tree-rules {1 [1 1] 0 [1 "[" 0 "]" 0]}]
      (is (= (state-to-string (nth-step binary-tree-rules 0 3))
             "1111[11[1[0]0]1[0]0]11[1[0]0]1[0]0"))))
  (testing "Using rules from strings"
    (let [binary-tree-rules {\1 (seq "11") \0 (seq "1[0]0")}]
      (is (= (state-to-string (nth-step binary-tree-rules \0 3))
             "1111[11[1[0]0]1[0]0]11[1[0]0]1[0]0")))))

(deftest test-step-with-productions
  (testing "3rd step of Fractal binary tree"
    (let* [binary-tree-rules {1 '(1 1) 0 '(1 "[" 0 "]" 0)}
           step (step-with-productions binary-tree-rules)]
      (is (= (state-to-string (->> 0 (step) (step) (step)))
             "1111[11[1[0]0]1[0]0]11[1[0]0]1[0]0")))))

(deftest test-bind-productions
  (testing "3rd step of Fractal (binary) tree using threading operators"
    (is (= (bind-productions {1 '(1 1) 0 '(1 "[" 0 "]" 0)}
                             (->> 0 (step) (step) (step) (state-to-string)))
           "1111[11[1[0]0]1[0]0]11[1[0]0]1[0]0")))
  (testing "3rd step of Fractal without threading operator")
  (is (= (state-to-string (bind-productions {1 '(1 1) 0 '(1 "[" 0 "]" 0)}
                          (step (step (step 0)))))
          "1111[11[1[0]0]1[0]0]11[1[0]0]1[0]0")))

test/lsystems/turtle_test.clj
(ns lsystems.turtle-test
  (:require [clojure.test :refer :all]
            [lsystems.turtle :refer :all]
            [lsystems.utils :refer :all]))

(deftest testing-turtle-movement
  (let [ps (new-pen-state 0 0)
        moved-once (forward ps 10)]
    (testing "Forward adds new line segment"
      (is (= (moved-once :lines) '({:from {:x 0 :y 0} :to {:x 0.0 :y -10.0}}))))
    (testing "Forward twice extends last line segment"
      (is (= ((forward moved-once 20) :lines) '({:from {:x 0 :y 0} :to {:x 0.0 :y -30.0}}))))
    (testing "Rotating then moving doesn't extend last line segment, but instead adds a new one"
      (let [moved-rotated-moved (forward (rotate moved-once 25) 10)]
        (is (= (count (moved-rotated-moved :lines)) 2))))
    (testing "Pushing state, moving, then popping resets to last position and angle but keeps the created lines"
      (let [pushed-moved-rotated-moved (forward (rotate (push-pos-and-angle moved-once) 25) 10)
            popped (pop-pos-and-angle pushed-moved-rotated-moved)]
        (is (= (count (pushed-moved-rotated-moved :stack)) 1))
        (is (= (count (popped :lines)) 2))
        (is (and (== (popped :x) 0) (== (popped :y) -10) (= (popped :facing) 0)))))
    (testing "If pen is up position should change but line segment should not be added"
      (let [moved (forward (pen-up ps) 10)]
        (is (empty? (moved :lines)))
        (is (and (== (moved :x) 0) (== (moved :y) -10)))
        (is (== ((forward moved 10) :y) -20))
        (let [pen-down-moved (forward (pen-down moved) 10)]
          (testing "If pen put back down it will draw a line"
            (is (== (pen-down-moved :y) -20))
            (is (= (pen-down-moved :lines) '({:from {:x 0.0 :y -10.0} :to {:x 0.0 :y -20.0}})))))))))

(deftest testing-execute-state-with-rules
  (let [state (seq "F+F") rules (standard-rule-set 10 90) ps (new-pen-state 0 0)]
    (testing "State is executed properly using the standard rule set"
      (let [executed (execute-state-with-rules state rules ps)]
        ;; TODO: check directions, I feel like this should be +10 not -10
        (is (and (approx-eq (executed :x) -10) (approx-eq (executed :y) -10)))
        (is (= (count (executed :lines)) 2))))
    (testing "Rules are called in order"
      (= (execute-state-with-rules
           (seq "123")
           {\1 (fn [s] (assoc s :lines (conj (s :lines) 1)))
            \2 (fn [s] (assoc s :lines (conj (s :lines) 2)))
            \3 (fn [s] (assoc s :lines (conj (s :lines) 3)))}
           ps)
         '(3 2 1))) ;; lists push to front, not to the back
    (testing "Missing rule uses identity instead"
      (= (execute-state-with-rules \I {} ps) ps))))


project.clj
(defproject lsystems "0.2.0-SNAPSHOT"
  :description "Small and simple L-systems implementation in Clojure"
  :url "https://github.com/joebentley/lsystems"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :plugins [[lein-codox "0.10.7"] [lein-environ "1.1.0"]]
  :dependencies [[org.clojure/clojure "1.10.0"] [clojure2d "1.1.0"] [environ "1.1.0"]]
  :repl-options {:init-ns lsystems.core}
  ;:main
  :profiles {:debug {:env {:debug? "true"}}}
  :codox {:source-uri "https://github.com/joebentley/lsystems/blob/master/{filepath}#L{line}"
          ;; /docs is the serve directory used by github pages
          :output-path "docs"
          ;; use MathJax so we can do LaTeX
          :html {:transforms [[:head] [:append
                                       [:script { :src "https://polyfill.io/v3/polyfill.min.js?features=es6"}]
                                       [:script { :id "MathJax-script" :async true :src "https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"}]]]}})


src/lsystems/core.clj
(ns lsystems.core
  (:require [clojure.walk :as walk]))

(defn keyword-name
  "Get keyword name without the colon"
  [keyword]
  (apply str (rest (str keyword))))

(defn combine-keywords
  "Takes multiple keywords and combines them into one, e.g. (combine-keywords :a :b) -> :ab"
  [& keywords]
  (if (empty? keywords)
    nil
    (keyword (apply str (map keyword-name keywords)))))

(defn string-to-keywords
  "Takes string of characters and turns them into keywords"
  [string]
  (map (comp keyword str) (seq string)))

(defn state-to-string
  "Convert list of keywords, numbers, strings to a single string"
  [keywords]
  (apply str (map (fn [x] (if (keyword? x) (keyword-name x) (str x))) keywords)))

(def new-productions
  "The productions are written as a hash map"
  (hash-map))

(defn add-production
  "Add new production from keyword to list of keywords"
  [productions from to]
  (assoc productions from to))

(defn step
  "Return next iteration of state using the given productions as a vector"
  [productions state]
  (if (not (vector? state))
    ;; wrap in vector if not already a vector
    (step productions (if (seq? state) (vec state)          ;; use vector instead of passed selection for performance
                                       (vector state)))     ;; otherwise just wrap in one
    (vec (flatten (map (fn [v] (get productions v v)) state)))))

(defn step-with-productions
  "Bind step function to given productions"
  [productions]
  (fn [state] (step productions state)))

(defmacro bind-productions
  "Bind productions in all following calls.
  Should probably use `(let [step (step-with-productions productions)] ...)` instead."
  [productions & exprs]
  `(do ~@(walk/prewalk (fn [expr] (if (and (list? expr) (= (first expr) 'step))
                                    `(step ~productions ~@(rest expr))
                                    expr))
                       exprs)))

(defn nth-step
  "Return the nth step using the productions and initial state"
  [productions state n]
  (let [step (step-with-productions productions)]
    (nth (iterate step state) n)))

src/lsystems/examples.clj
(ns lsystems.examples
  (:require
    [lsystems.core :refer [nth-step]]
    [lsystems.turtle :refer :all]
    [clojure2d.core :as c]
    [clojure2d.color :as col]))

(defn fractal-binary-tree []
  (render-to-canvas-by-executing-state-with-rules
    (nth-step {1 '(1 1) 0 '(1 "[" 0 "]" 0)} 0 9)
    {0 (fn [s] (forward s 0.2))
     1 (fn [s] (forward s 0.4))
     "[" (fn [s] (-> s (push-pos-and-angle) (rotate (- 45))))
     "]" (fn [s] (-> s (pop-pos-and-angle) (rotate (+ 45))))}

    (fn [canvas]
      (show-window canvas "Fractal binary tree")
      (save-canvas-to-file canvas "example-renders/fractal-binary-tree.jpg"))
    :auto-resize-padding 50))

(defn fractal-plant []
  (let [width 600 height 600]
    (render-to-canvas-by-executing-state-with-rules
      (nth-step {\X (seq "F+[[X]-X]-F[-FX]+X") \F '(\F \F)}
                \X
                6)
      (standard-rule-set 10 25)

      (fn [canvas]
        (show-window canvas "Fractal plant")
        (save-canvas-to-file canvas "example-renders/fractal-plant.jpg"))

      :facing -25 :width width :height height :auto-resize-padding 50
      :canvas-function (fn [canvas]
                         (c/gradient-mode canvas 0 0 :white 0 height :light-green)
                         (c/rect canvas 0 0 width height)
                         (c/set-color canvas :forest-green)
                         (c/set-stroke canvas 1.3)))))

(defn dragon-curve []
  (render-to-canvas-by-executing-state-with-rules
    (nth-step {\X (seq "X+YF+") \Y (seq "-FX-Y")}
              '(\F\X)
              10)
    {\F (fn [s] (forward s 10))
     \- (fn [s] (rotate s (- 90)))
     \+ (fn [s] (rotate s (+ 90)))}

    (fn [canvas]
      (show-window canvas "Dragon curve")
      (save-canvas-to-file canvas "example-renders/dragon-curve.jpg"))

    :height 400 :auto-resize-padding 50
    :canvas-function (fn [canvas]
                       (c/set-background canvas :skyblue)
                       (c/set-color canvas :white)
                       (c/set-stroke canvas 2))))

(defn abop-fig1.24 []
  "Example of using render-to-canvas-grid

  Source: http://algorithmicbotany.org/papers/abop/abop-ch1.pdf"
  (let [palette (take-last 6 (col/resample 10 (col/palette-presets :greens-6)))
        width 2000 height 2000]
    (render-to-canvas-grid
      3 2 width height

      (fn [canvas]
        ;(show-window canvas "abop-fig1.24")
        (save-canvas-to-file canvas "example-renders/abop-fig1.24.png"))

      (map-indexed
           ;; set color from palette for each figure
           (fn [i m] (assoc m :canvas-function (fn [canvas] (c/set-color canvas (nth palette i)))))
           (list
             {:state (nth-step {\F (seq "F[+F]F[-F]F")} \F 5)
              :rules (standard-rule-set 10 25.7)}
             {:state (nth-step {\F (seq "F[+F]F[-F][F]")} \F 5)
              :rules (standard-rule-set 10 20)}
             {:state (nth-step {\F (seq "FF-[-F+F+F]+[+F-F-F]")} \F 4)
              :rules (standard-rule-set 10 22.5)}
             {:state (nth-step {\X (seq "F[+X]F[-X]+X") \F '(\F \F)} \X 7)
              :rules (standard-rule-set 10 20)}
             {:state (nth-step {\X (seq "F[+X][-X]FX") \F '(\F \F)} \X 7)
              :rules (standard-rule-set 10 25.7)}
             {:state (nth-step {\X (seq "F-[[X]+X]+F[+FX]-X") \F '(\F \F)} \X 5)
              :rules (standard-rule-set 10 22.5)}))

      :canvas-function (fn [canvas]
                         (c/gradient-mode canvas 0 0 :white width height :light-blue)
                         (c/rect canvas 0 0 width height)
                         (c/set-stroke canvas 3))
      :padding 50)))

(def all-examples [#'fractal-binary-tree #'fractal-plant #'dragon-curve #'abop-fig1.24])

(defn -main
  "Parse arguments and run the given example"
  [& args]
  (let [fn-name (fn [f] (:name (meta f)))]
    (if (or (empty? args) (some #{"--list"} args))          ;; is argument list empty or contains "--list"?
      ;; list all the examples
      (do (println "examples:")
          (doseq [example all-examples]
            (println (fn-name example))))
      (if (some #{"--all"} args)
        ;; do all of the examples
        (doseq [example all-examples] (example))
        ;; get the example
        (let [chosen-example (last args)
              index (.indexOf (map (comp str fn-name) all-examples) chosen-example)]
          (if-not (neg? index)
            (do (println (str "running example: " chosen-example "..."))
                ((all-examples index)))
            (println (str "example " chosen-example " not found"))))))))